### ISSUE TITLE
Shrinking values for small constant integers during store operation

### ISSUE THREAD

**Issue by MaxGraey** [WebAssembly/binaryen#3077]
Found one size reduction possibility for constants in load operations.

Currently this code:
```wat
(func $a (param i32)
  local.get 0
  i32.const 2147483647
  i32.store8
)
```
takes `11` bytes

but could be optimized to
```wat
(func $a (param i32)
  local.get 0
  i32.const 255
  i32.store8
)
```
which takes only `8` bytes.

We could do similar optimizations for `i32.store(8|16)` and `i64.store(8|16|32)`

```wat
i32.store8(dst, C)   ->   i32.store8(dst, C & 0xFF)
i32.store16(dst, C)  ->   i32.store16(dst, C & 0xFFFF)

i64.store8(dst, C)   ->   i64.store8(dst, C & 0xFF)
i64.store16(dst, C)  ->   i64.store16(dst, C & 0xFFFF)
i64.store32(dst, C)  ->   i64.store32(dst, C & 0xFFFFFFFF) 
```

~~However not sure it possible for any constant values. Perhaps only for positives~~

**Comment by MaxGraey** [comment]
up. Does it make sense?

**Comment by binji** [comment]
Yep, makes sense to me. You could do a similar thing (if it isn't already there) for shift/rotate instructions, since those mask the shift amount.

**Comment by MaxGraey** [comment]
> You could do a similar thing (if it isn't already there) for shift/rotate instructions

Yeah, also thinking about this reduction in #3099. But not yet implemented. Perhaps I'll do this in separate PR.
Will be great if someone help me with review. At the moment, there are about pending 20 PRs =)

**Comment by tlively** [comment]
@MaxGraey I'll start working through the review backlog 👍

**Comment by MaxGraey** [comment]
@tlively That would be awesome! Thanks!

**Comment by MaxGraey** [comment]
Implemented in #3117
